keras_nlp/layers/fnet_encoder.py

```diff
@@ -0,0 +1,157 @@
+            normalization components.
+        name: string, defaults to None. The name of the layer.
+        **kwargs: other keyword arguments.
+
+    Examples:
+
+    ```python
+    # Create a single FNet encoder layer.
+    encoder = keras_nlp.layers.FNetEncoder(
+        intermediate_dim=64)
+
+    # Create a simple model containing the encoder.
+    input = tf.keras.Input(shape=[4, 6])
+    output = encoder(input)
+    model = tf.keras.Model(inputs=input, outputs=output)
+
+    # Call encoder on the inputs.
+    input_data = tf.random.uniform(shape=[1, 10, 64])
+    output = model(input_data)
+
+    ```
+
+    References:
+        [Lee-Thorp et al., 2021](https://arxiv.org/abs/2105.03824)
+    """
+
+    def __init__(
+        self,
+        intermediate_dim=3072,
+        dropout=0.1,
+        activation="gelu",
+        layer_norm_epsilon=1e-12,
+        name=None,
+        **kwargs
+    ):
+        super().__init__(name=name, **kwargs)
+        self.intermediate_dim = intermediate_dim
+        self.dropout = dropout
+        self.activation = activation
+        self.layer_norm_epsilon = layer_norm_epsilon
+        self._built = False
+
+    def _build(self, input_shape):
+        # Create layers based on input shape.
+        self._built = True
+        feature_size = input_shape[-1]
+
+        # Layer Norm layers.
+        self._mixing_layer_norm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon
+        )
+        self._output_layer_norm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon
+        )
+
+        # Feedforward layer.
+        self._intermediate_dense = keras.layers.Dense(
+            self.intermediate_dim, activation=self.activation
+        )
+        self._output_dense = keras.layers.Dense(feature_size)
+        self._output_dropout = keras.layers.Dropout(rate=self.dropout)
+
+    def _fourier_transform(self, input):
+        # Apply FFT on the input and take the real part. Before we apply fourier
+        # transform, let's convert the dtype of the input tensor to complex64.
+        input = tf.cast(input, tf.complex64)
+        mixing_output = tf.math.real(tf.signal.fft2d(input))
+        return mixing_output
+
+    def _add_and_norm(self, input1, input2, norm_layer):
+        return norm_layer(input1 + input2)
+
+    def _feed_forward(self, input):
+        x = self._intermediate_dense(input)
+        x = self._output_dense(x)
+        return self._output_dropout(x)
+
+    def call(self, inputs):
+        """Forward pass of the FNetEncoder.
+
+        Args:
+            inputs: a Tensor. The input data to TransformerEncoder, should be
+                of shape [batch_size, sequence_length, feature_dim].
+
+        Returns:
+            A Tensor of the same shape as the `inputs`.
+        """
+
+        if not self._built:
+            self._build(inputs.shape)
+
+        # Apply fourier transform on the input. Note: We don't have padding
```

Comment:
I want to understand this a bit more - are you saying the original paper does not do padding, and take in RaggedTensor as input? Thanks!

Reply:
So, as far as I understand, they do pad the sequences so that every sequence is of the same length. See this: https://github.com/google-research/google-research/blob/master/f_net/layers.py#L106. The expected shape is `(bsz, max_seq_length, hidden_dim)`.

However, the padding mask is not used at all. They delete it here: https://github.com/google-research/google-research/blob/master/f_net/layers.py#L114. So, the mixing is done over all the tokens, including the padding tokens. See the input pipeline here: https://github.com/google-research/google-research/blob/master/f_net/input_pipeline.py#L100.

Reply:
I received a reply from the first author, James Lee-Thorp. This is the gist of the message:

**For FNet, padding tokens are applied, but padding tokens are not masked**. There is a possibility that the model may become sensitive to samples with varying number of padding tokens...but in practice, the authors did not notice any major performance issues. So, I think we can keep it the same, we don't need to apply padding masks.

If we do want to apply a padding mask, we should only apply it after the embedding layer, and not after every Fourier sublayer. This is because if we apply it after every Fourier sublayer, we will end up zeroing out some frequencies which is counter-productive.